File: frontend/src/services/participationService.js
import api from './apiService';

export const participationService = {
  // Get all participations with pagination
  getAllParticipations: async (page = 0, size = 20, sort = 'participationDate,desc') => {
    const response = await api.get('/admin/participations', {
      params: { page, size, sort }
    });
    return response.data;
  },

  // Get participation by ID
  getParticipationById: async (id) => {
    const response = await api.get(`/admin/participations/${id}`);
    return response.data;
  },

  // Get participations by semester with pagination
  getParticipationsBySemester: async (semesterId, page = 0, size = 20) => {
    const response = await api.get(`/admin/participations/semester/${semesterId}`, {
      params: { page, size }
    });
    return response.data;
  },

  // Get participations by group with pagination
  getParticipationsByGroup: async (groupId, page = 0, size = 20) => {
    const response = await api.get(`/admin/participations/group/${groupId}`, {
      params: { page, size }
    });
    return response.data;
  },

  // Get participations by person with pagination
  getParticipationsByPerson: async (personId, page = 0, size = 20) => {
    const response = await api.get(`/admin/participations/person/${personId}`, {
      params: { page, size }
    });
    return response.data;
  },

  // Search participations with filters and pagination
  searchParticipations: async (filters, page = 0, size = 20, sort = 'participationDate,desc') => {
    const response = await api.get('/admin/participations/search', {
      params: { ...filters, page, size, sort }
    });
    return response.data;
  },

  // Create participation
  createParticipation: async (semesterId, groupId, personId, participationData) => {
    const response = await api.post(
      `/admin/participations?semesterId=${semesterId}&groupId=${groupId}&personId=${personId}`,
      participationData
    );
    return response.data;
  },

  // Update participation
  updateParticipation: async (id, participationData) => {
    const response = await api.put(`/admin/participations/${id}`, participationData);
    return response.data;
  },

  // Delete participation
  deleteParticipation: async (id) => {
    const response = await api.delete(`/admin/participations/${id}`);
    return response.data;
  }
};

File: src/main/java/com/example/sample/service/ParticipationService.java
package com.example.sample.service;

import com.example.sample.model.Group;
import com.example.sample.model.Participation;
import com.example.sample.model.Person;
import com.example.sample.model.Semester;
import com.example.sample.repository.GroupRepository;
import com.example.sample.repository.ParticipationRepository;
import com.example.sample.repository.PersonRepository;
import com.example.sample.repository.SemesterRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class ParticipationService {

    private final ParticipationRepository participationRepository;
    private final SemesterRepository semesterRepository;
    private final GroupRepository groupRepository;
    private final PersonRepository personRepository;

    @Autowired
    public ParticipationService(ParticipationRepository participationRepository,
                               SemesterRepository semesterRepository,
                               GroupRepository groupRepository,
                               PersonRepository personRepository) {
        this.participationRepository = participationRepository;
        this.semesterRepository = semesterRepository;
        this.groupRepository = groupRepository;
        this.personRepository = personRepository;
    }

    public List<Participation> getAllParticipations() {
        return participationRepository.findAll();
    }

    public Optional<Participation> getParticipationById(Integer id) {
        return participationRepository.findById(id);
    }

    public List<Participation> getParticipationsBySemester(Integer semesterId) {
        return participationRepository.findBySemester_SemesterId(semesterId);
    }

    public List<Participation> getParticipationsByGroup(Integer groupId) {
        return participationRepository.findByGroup_GroupId(groupId);
    }

    public List<Participation> getParticipationsByPerson(Integer personId) {
        return participationRepository.findByPerson_PersonId(personId);
    }

    public Participation createParticipation(Participation participation,
                                            Integer semesterId,
                                            Integer groupId,
                                            Integer personId) {
        Semester semester = semesterRepository.findById(semesterId)
                .orElseThrow(() -> new IllegalArgumentException("Semester not found with id: " + semesterId));

        Group group = groupRepository.findById(groupId)
                .orElseThrow(() -> new IllegalArgumentException("Group not found with id: " + groupId));

        Person person = personRepository.findById(personId)
                .orElseThrow(() -> new IllegalArgumentException("Person not found with id: " + personId));

        String status = participation.getStatus().toLowerCase();
        if (!status.equals("ontime") && !status.equals("late") && !status.equals("absent")) {
            throw new IllegalArgumentException("Status must be 'ontime', 'late', or 'absent'");
        }

        if (participation.getParticipationDate() == null) {
            throw new IllegalArgumentException("Participation date is required");
        }

        participation.setSemester(semester);
        participation.setGroup(group);
        participation.setPerson(person);
        return participationRepository.save(participation);
    }

    public Participation updateParticipation(Integer id, Participation participationDetails) {
        Participation participation = participationRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Participation not found with id: " + id));

        if (participationDetails.getStatus() != null) {
            String status = participationDetails.getStatus().toLowerCase();
            if (!status.equals("ontime") && !status.equals("late") && !status.equals("absent")) {
                throw new IllegalArgumentException("Status must be 'ontime', 'late', or 'absent'");
            }
            participation.setStatus(participationDetails.getStatus());
        }

        if (participationDetails.getParticipationDate() != null) {
            participation.setParticipationDate(participationDetails.getParticipationDate());
        }

        return participationRepository.save(participation);
    }

    public void deleteParticipation(Integer id) {
        if (!participationRepository.existsById(id)) {
            throw new IllegalArgumentException("Participation not found with id: " + id);
        }
        participationRepository.deleteById(id);
    }

    // Pagination methods
    public Page<Participation> getAllParticipationsPaginated(Pageable pageable) {
        return participationRepository.findAll(pageable);
    }

    public Page<Participation> getParticipationsBySemesterPaginated(Integer semesterId, Pageable pageable) {
        return participationRepository.findBySemester_SemesterId(semesterId, pageable);
    }

    public Page<Participation> getParticipationsByGroupPaginated(Integer groupId, Pageable pageable) {
        return participationRepository.findByGroup_GroupId(groupId, pageable);
    }

    public Page<Participation> getParticipationsByPersonPaginated(Integer personId, Pageable pageable) {
        return participationRepository.findByPerson_PersonId(personId, pageable);
    }

    public Page<Participation> searchParticipations(Integer semesterId, Integer groupId, Integer personId,
                                                     String status, LocalDate startDate, LocalDate endDate,
                                                     Pageable pageable) {
        Specification<Participation> spec = Specification.where(null);

        if (semesterId != null) {
            spec = spec.and((root, query, cb) ->
                cb.equal(root.get("semester").get("semesterId"), semesterId));
        }
        if (groupId != null) {
            spec = spec.and((root, query, cb) ->
                cb.equal(root.get("group").get("groupId"), groupId));
        }
        if (personId != null) {
            spec = spec.and((root, query, cb) ->
                cb.equal(root.get("person").get("personId"), personId));
        }
        if (status != null && !status.isEmpty()) {
            spec = spec.and((root, query, cb) ->
                cb.equal(cb.lower(root.get("status")), status.toLowerCase()));
        }
        if (startDate != null) {
            spec = spec.and((root, query, cb) ->
                cb.greaterThanOrEqualTo(root.get("participationDate"), startDate));
        }
        if (endDate != null) {
            spec = spec.and((root, query, cb) ->
                cb.lessThanOrEqualTo(root.get("participationDate"), endDate));
        }

        return participationRepository.findAll(spec, pageable);
    }
}